### ISSUE TITLE
Infinite recursion possible in OPA v0.55.0 and below

### ISSUE THREAD

**Issue by adam-sandor** [open-policy-agent/opa#6316]
## Short description 

Playground reproduction: https://play.openpolicyagent.org/p/LjU9gPo507

Running this same policy in OPA 0.55.0 doesn't produce the same `rego_recursion_error`s. It will complete successfully with the expected output. This tells me that there has been a regression between 0.55.0 and 0.56.0 versions in either the recursion detection or the actual evaluation. 

OPA version: 0.55.0 - works as expected, 0.56.0 & 0.57.0 produces `rego_recursion_error`

## Expected behavior
Successful evaluation

**Comment by adam-sandor** [comment]
This change to the rules fixes it: https://play.openpolicyagent.org/p/EQQ4PaHsy0

I moved the "rule_sets" into a sub-object under the key "rule_sets" in results. This way the final_allow rule is not iterating over keys in results. Even though previously it was only iterating over keys that were safe to use in that context it seems that 0.56.0 is picking that up as a recursion_error. I think the rules are actually better after the workaround so this problem doesn't affect me. It's still an unexpected change in behaviour between minor versions so you might want to check that out.

**Comment by anderseknert** [comment]
Hey Adam! Good to have you around 😃 This is a great observation 👍 However, I don't think successful evaluation should be the expected behavior in this case. Since the keys are provided as dynamic data, the compiler can't know whether they are safe or not, and if it doesn't, it _has_ to bail out. Trying this out on OPA v0.55.0 as you suggested, and changing the  `rule_sets` array in your dataset to include "result":
```json
"rule_sets": [
    "actors",
    "movies",
    "result"
]
```
...and for the first time, I'm now seeing infinte recursion actually happen in OPA 😱

Given how this is fixed in recent versions, I'm not sure there's anything actionable wrt current code.
Drafting a CVE for v0.55.0 and below seems like it might be a reasonable course of action though... 😬

**Comment by adam-sandor** [comment]
Sounds good to me!

**Comment by anderseknert** [comment]
Perhaps it can be reduced down further, but here's a fairly minimal example to reproduce using OPA v0.55.0 and below:

**p.rego**
```rego
package foo

results[id] := {"result": result[id]} {
	id := input.foo.bar[_]
}

results["result"] := "allow" {
	final_allow
}

final_allow {
    id := input.foo.bar[_]
    print(id)
    results[id].result == "deny"
}

result[input.foo.bar[_]] := "allow" {
	false
}
```
**input.json**
```json
{
    "foo": {
        "bar": [
            "result"
        ]
    }
}
```

```
$ opa eval -d p2.rego -i input.json data.foo
result
result
result
result
result
result
result
...
thousands of lines later
...
result
result
result
result
result
result
result
runtime: goroutine stack exceeds 1000000000-byte limit
runtime: sp=0x14020602340 stack=[0x14020602000, 0x14040602000]
fatal error: stack overflow

runtime stack:
runtime.throw({0x100f0285d?, 0x1004fc6f8?})
	/opt/homebrew/Cellar/go/1.21.2/libexec/src/runtime/panic.go:1077 +0x40 fp=0x171146d50 sp=0x171146d20 pc=0x1005066a0
runtime.newstack()
	/opt/homebrew/Cellar/go/1.21.2/libexec/src/runtime/stack.go:1107 +0x458 fp=0x171146f00 sp=0x171146d50 pc=0x100521ba8
traceback: unexpected SPWRITE function runtime.morestack
runtime.morestack()
	/opt/homebrew/Cellar/go/1.21.2/libexec/src/runtime/asm_arm64.s:316 +0x70 fp=0x171146f00 sp=0x171146f00 pc=0x10053bb30

goroutine 1 [running]:
strconv.ParseUint({0x100f21f2e?, 0x1?}, 0xa?, 0x40?)
```

**Comment by anderseknert** [comment]
Given that this affects older versions of OPA, and that triggering this condition requires quite some fiddling with the policy — i.e. can't be triggered by just any arbitrary input — perhaps a good enough resolution to this would be to ensure there's a test in place covering the example provided above. If this was "accidentally fixed" before, we should at least make sure it's not accidentally unfixed in the future.

**Comment by srenatus** [comment]
Agreed! Bonus points for bisecting the commit that fixed it. My bet is on @johanfylling 😉

**Comment by johanfylling** [comment]
I think I know what's happening here 🤔 . 

TL;DR: when adding support for general refs in rule heads, we tweaked how the rule tree is organized. In earlier versions of OPA, it could be that some rules can become "hidden" to the recursion-checker.

We can simplify the policy further:

```rego
package recursion

# R1
results[id] := 2 {
  id := "bar"
}

# R2
results.foo := 1 {
  final_allow
}

# R3
final_allow {
  id := "foo"
  print(id)
  results[id] == 3
}
```

The above policy will trigger a `rego_recursion_error` in OPA `0.56.0`, and the infinite recursion bug in older versions. 
I think in older versions, `R2` would be positioned as a child to `R1` in the rule tree; whereas in newer versions, they'll be placed on the same node as siblings. The hypothesis is that, once a rule is found, the recursion-checker will not check for rules deeper in the tree, and miss `R2`. The recursion happens between `R2` and `R3`, but `R1` is required to trick the checker.
This is a qualified guess, though, so I could be wrong.

**Comment by johanfylling** [comment]
PR with regression test: https://github.com/open-policy-agent/opa/pull/6318

**Comment by johanfylling** [comment]
Regression test added. Closing this issue. Thank you for reporting it! 👍 😃
